src/main/java/net/greghaines/jesque/worker/WorkerImpl.java

```
@@ -397,46 +394,44 @@ protected int getReconnectAttempts() {
-        int missCount = 0;
+        int currentPriority = 0;
         String curQueue = null;
-        while (RUNNING.equals(this.state.get())) {
-            try {
+        try {
+            for (;RUNNING.equals(this.state.get());checkPaused()) {
+                curQueue = getCurrentQueue(currentPriority);
                 if (threadNameChangingEnabled) {
                     renameThread("Waiting for " + JesqueUtils.join(",", this.queueNames));
                 }
-                curQueue = this.queueNames.poll(EMPTY_QUEUE_SLEEP_TIME, TimeUnit.MILLISECONDS);
-                if (curQueue != null) {
-                    this.queueNames.add(curQueue); // Rotate the queues
-                    checkPaused(); 
-                    // Might have been waiting in poll()/checkPaused() for a while
-                    if (RUNNING.equals(this.state.get())) {
-                        this.listenerDelegate.fireEvent(WORKER_POLL, this, curQueue, null, null, null, null);
-                        final String payload = pop(curQueue);
-                        if (payload != null) {
-                            final Job job = ObjectMapperFactory.get().readValue(payload, Job.class);
-                            process(job, curQueue);
-                            missCount = 0;
-                        } else if (++missCount >= this.queueNames.size() && RUNNING.equals(this.state.get())) {
-                            // Keeps worker from busy-spinning on empty queues
-                            missCount = 0;
-                            Thread.sleep(EMPTY_QUEUE_SLEEP_TIME);
-                        }
+                final String payload = pop(curQueue);
+                if (payload != null) {
+                    currentPriority = 0;
+                    process(ObjectMapperFactory.get().readValue(payload, Job.class), curQueue);
+                } else {
+                    currentPriority++;
+                    if(currentPriority == queueNames.size()) {
```

right - so I'd make the list immutable and/or throw an `UnsupportedOperation` exception for Q removal
(and document it via tests)